app/src/common/shared/com/igalia/wolvic/search/SearchEngineWrapper.kt

```diff
@@ -0,0 +1,231 @@
+            if (intent.action == Intent.ACTION_LOCALE_CHANGED) {
+                val userSearchEngineId = SettingsStore.getInstance(context).searchEngineId
+                setupSearchEngine(context, userSearchEngineId)
+            }
+        }
+    }
+
+    init {
+        mContextRef = WeakReference(aContext)
+        mPrefs = PreferenceManager.getDefaultSharedPreferences(context!!)
+        mAutocompleteEnabled = SettingsStore.getInstance(context!!).isAutocompleteEnabled
+        val preferredSearchEngineId = SettingsStore.getInstance(context!!).searchEngineId
+        setupSearchEngine(aContext, preferredSearchEngineId)
+    }
+
+    val availableSearchEngines: Collection<SearchEngine>
+        get() {
+            updateSearchEngine()
+            return mSearchEnginesMap!!.values
+        }
+
+    fun setDefaultSearchEngine() {
+        if (hasContext()) setupSearchEngine(context!!, null)
+    }
+
+    fun setCurrentSearchEngineId(context: Context?, searchEngineId: String?) {
+        if (hasContext()) SettingsStore.getInstance(context!!).searchEngineId = searchEngineId
+    }
+
+    /**
+     * We cannot send system ACTION_LOCALE_CHANGED when the component refreshes the engines
+     * with the updated information so we have to update the whole manager.
+     *
+     * @param aContext Activity context
+     * @param userPref User preferred engine (among the available ones)
+     */
+    private fun setupSearchEngine(aContext: Context?, userPref: String?) {
+        val thisContext: Context = aContext ?: context!!
+
+        val ioDispatcher: CoroutineContext = Dispatchers.IO
+        val browserStore = BrowserStore(
+            middleware = listOf(
+                SearchMiddleware(thisContext,
+                    ioDispatcher = ioDispatcher)
+            ))
+
+        val data = GeolocationData.parse(SettingsStore.getInstance(thisContext).geolocationData)
+        val regionState = if (data != null) {
+            Log.d(LOGTAG, "Using Geolocation based search localization: $data")
+            // If we have geolocation data we initialize with the received data.
+            val country = data.countryCode
+            val language = Locale.getDefault().language
+            RegionState(language, country)
+        } else {
+            val locale = thisContext.resources.configuration.locales.get(0)
+            RegionState(locale.language, locale.country)
+        }
+
+        browserStore.dispatch(
+            SearchAction.SetRegionAction(
+                regionState,
+            ),
+        )
+
+        // Since we have async state loads, let's wait for some time for it to get loaded.
+        val time = System.currentTimeMillis()
+        while (browserStore.state.search.searchEngines.isEmpty()) {
+            if (System.currentTimeMillis() - time > 5000) {
+                throw NullPointerException("searchEngines list is still empty after timeout exceeds")
+            }
+        }
```

Comment:
> Although the code compiles, now we can't get any of the search engines. We must miss something here.

It turned out to be that the code now executes asynchronously and we must wait for some time before it's available. I know this is something really dirty here, but I'm not sure if we have other options.

Reply:
Isn't there any event we can listen to or a callback we can pass as argument, that is executed when the operation is completed ?

Reply:
I’m not aware of any event or callback here, but even if we have those things, we will still need to block and wait to make it synchronised.

Reply:
Sorry, I wasn't following this close enough, so my question may be dumb or obvious, but why we need to block and wait ? Couldn't we just render an empty list until we receive the event or the callback we may pass is called, and only then rendering an updated list of search engines ? 

Obviously all this under the assumption that there is an event or callback that we can use.

Reply:
This method is called when wolvic starts, and we need a valid one so that SearchSuggestionClient can be properly initialised.

Reply:
Finally get this dirty thing removed, without using any callback or block, just check every time when the variables are fetched / methods are called and continue the unfinished init.